flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/OutputColsHelper.java

```diff
@@ -0,0 +1,190 @@
+				reservedColIndices.add(i);
+				reservedColToResultIndex.add(index++);
+			}
+		}
+		for (int i = 0; i < outputToResultIndices.length; i++) {
+			if (outputToResultIndices[i] == -1) {
+				outputToResultIndices[i] = index++;
+			}
+		}
+		this.resultLength = index;
+		this.reservedColIndices = new int[reservedColIndices.size()];
+		this.reservedToResultIndices = new int[reservedColIndices.size()];
+		for (int i = 0; i < this.reservedColIndices.length; i++) {
+			this.reservedColIndices[i] = reservedColIndices.get(i);
+			this.reservedToResultIndices[i] = reservedColToResultIndex.get(i);
+		}
+	}
+
+	/**
+	 * Get the reserved colNames, the result is [reserve columns] subtract [output columns].
+	 *
+	 * @return the reserved colNames.
+	 */
+	public String[] getReservedColNames() {
+		String[] reservedColNames = new String[reservedColIndices.length];
+		for (int i = 0; i < reservedColIndices.length; i++) {
+			reservedColNames[i] = dataColNames[reservedColIndices[i]];
+		}
+		return reservedColNames;
+	}
+
+	/**
+	 * Get the final table schema. [result columns] = ([reserve columns] subtract [output columns]) + [output columns]
+	 *
+	 * @return the table schema.
+	 */
+	public TableSchema getResultSchema() {
+		String[] resultColNames = new String[resultLength];
+		TypeInformation[] resultColTypes = new TypeInformation[resultLength];
+		for (int i = 0; i < reservedColIndices.length; i++) {
+			resultColNames[reservedToResultIndices[i]] = dataColNames[reservedColIndices[i]];
+			resultColTypes[reservedToResultIndices[i]] = dataColTypes[reservedColIndices[i]];
+		}
+		for (int i = 0; i < outputToResultIndices.length; i++) {
+			resultColNames[outputToResultIndices[i]] = outputColNames[i];
+			resultColTypes[outputToResultIndices[i]] = outputColTypes[i];
+		}
+		return new TableSchema(resultColNames, resultColTypes);
+	}
+
+	/**
+	 * Get the outputting row, which is a combination of input data and output data.
+	 *
+	 * @param data   input data
+	 * @param output output data
+	 * @return The outputting row, which is a combination of reserved input data and prediction result.
+	 */
+	public Row getResultRow(Row data, Row output) {
+		int numOutputs = outputToResultIndices.length;
+		if (output.getArity() != numOutputs) {
+			throw new IllegalArgumentException("Invalid output size");
+		}
+		Row result = new Row(resultLength);
+		for (int i = 0; i < reservedColIndices.length; i++) {
+			result.setField(reservedToResultIndices[i], data.getField(reservedColIndices[i]));
+		}
+		for (int i = 0; i < numOutputs; i++) {
+			result.setField(outputToResultIndices[i], output.getField(i));
+		}
+		return result;
+	}
+
+	/**
+	 * Get the outputting row, which is a combination of input data and output data.
+	 * The difference with getResultRow is that it's single column output.
+	 *
+	 * @param data   input data
+	 * @param output output data
+	 * @return
+	 */
+	public Row getResultRowSingle(Row data, Object output) {
```

Comment:
do we really need this API? it seems like the following are identical
```
getResultRowSingle(someData, myObj);
getResultRow(someData, Row.of(myObj));
```

Reply:
Thanks, I removed the method getResultRowSingle().